libdevice/sanitizer_utils.cpp

```diff
@@ -116,22 +116,44 @@ inline uptr MemToShadow_CPU(uptr addr) {
 }
 
 inline uptr MemToShadow_DG2(uptr addr, uint32_t as) {
-  uptr shadow_ptr = 0;
-  if (addr & (~0xffffffffffff)) {
-    shadow_ptr =
-        (((addr & 0xffffffffffff) >> 3) + __AsanShadowMemoryGlobalStart) |
-        (~0xffffffffffff);
-  } else {
-    shadow_ptr = (addr >> 3) + __AsanShadowMemoryGlobalStart;
+  if (as == ADDRESS_SPACE_GENERIC) {
+    auto old = addr;
+    if ((addr = (uptr)ToPrivate((void *)old))) {
+      as = ADDRESS_SPACE_PRIVATE;
+    } else if ((addr = (uptr)ToLocal((void *)old))) {
+      as = ADDRESS_SPACE_LOCAL;
+    } else if ((addr = (uptr)ToGlobal((void *)old))) {
+      as = ADDRESS_SPACE_GLOBAL;
+    } else {
+      if (__AsanDebug)
+        __spirv_ocl_printf(__generic_to_fail, old);
+      return 0;
+    }
+    if (__AsanDebug)
+      __spirv_ocl_printf(__generic_to, old, addr, as);
   }
```

Comment:
NIT: I think this should be outlined into some helper function as the same code is used in MemToShadow_DG2 and most likely will be used for other hardware in the future.

Reply:
Good idea! Done